### PR TITLE
Introduce test for use case where AllocateByIp fails with t2

### PR DESCRIPTION
**Pull request by flashvoid**

### CODE REVIEW

File: common/defs.go [hunk 1]
```diff
@@ -21,10 +21,19 @@ import (
 	"fmt"
 	"net"
 	"strings"
+	"encoding/json"
 )
 
 // Here we only keep type definitions and struct definitions with no behavior.
 
+func String(i interface{}) string {
```
Comment (jbrendel) [inline common/defs.go]: Is 'String' really the most descriptive name we can give this function? The name is very ... generic and doesn't really tell us what it does.

File: common/defs.go [hunk 2]
```diff
@@ -99,6 +108,9 @@ type PortUpdateMessage struct {
 	Port uint64 `json:"port"`
 }
 
+// Endpoint represents an endpoint - that is, something that
```
Comment (jbrendel) [inline common/defs.go]: I like this docstring. That's the sort of useful, descriptive docstring I'd like to see more of.

File: common/defs.go [hunk 3]
```diff
@@ -211,7 +235,7 @@ func validateRules(rules Rules) []string {
 			errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid protocol: %s.", ruleNo, r.Protocol))
 		}
 		if r.Protocol == "tcp" {
-			r.IsStateful = true
+			rules[i].IsStateful = true
```
Comment (jbrendel) [inline common/defs.go]: Why was this change necessary? Shouldn't "r" here be the same as "rules[i]" ?

Comment (cgilmour) [inline common/defs.go]: `r` is a copy of the value at `rules[i]`. To change the original, you need to access the original via the slice (`rules`) and index(`i`).

File: common/defs.go (hunk 1 above)
Comment (cgilmour) [inline common/defs.go]: It's used by other types that provide a `String() string` method (to satisfy the `fmt.Stringer` interface) -- it's a convenience function to avoid writing repetitive code in those methods.

I think it's a bit odd; produce JSON if possible otherwise emit the default conversion to string.
That doesn't seem helpful for logging, troubleshooting, etc that it would affect.

File: ipam/ipam.go [hunk 4]
```diff
@@ -79,15 +79,21 @@ func (ipam *IPAM) allocateIP(input interface{}, ctx common.RestContext) (interfa
 
 	// check for missing/empty required parameters
 	if tenantParam == "" {
-		return nil, errors.New("Missing or empty tenantName/tenantID parameter")
+		err := errors.New("Missing or empty tenantName/tenantID parameter")
+		log.Printf("IPAM encountered an error: %v", err)
```
Comment (cgilmour) [inline ipam/ipam.go]: Why log it here instead of the caller who has (or should) to log and handle the error anyway?

File: ipam/ipam.go [hunk 5]
```diff
@@ -169,26 +182,33 @@ func (ipam *IPAM) allocateIP(input interface{}, ctx common.RestContext) (interfa
...
 
 	segmentsUrl := fmt.Sprintf("/tenants/%s/segments", endpoint.TenantID)
 	var segments []tenant.Segment
 	err = client.Get(segmentsUrl, &segments)
 	if err != nil {
+		log.Printf("IPAM encountered an error: %v", err)
 		return nil, err
 	}
 	found = false
+	fmt.Printf("IPAM found %d segments for tenant %s\n", len(segments), endpoint.TenantID)
```
Comment (cgilmour) [inline ipam/ipam.go]: Why `fmt` instead of `log`?

File: ipam/ipam.go [hunk 6]
```diff
@@ -198,78 +218,87 @@ func (ipam *IPAM) allocateIP(input interface{}, ctx common.RestContext) (interfa
...
 	log.Printf("Parsing Romana IP address of host %s: %s\n", host.Name, host.RomanaIp)
 	_, network, err := net.ParseCIDR(host.RomanaIp)
 	if err != nil {
+		log.Printf("IPAM encountered an error parsing %s: %v", host.RomanaIp, err)
 		return nil, err
 	}
 	hostIpInt := common.IPv4ToInt(network.IP)
-	tSeq := t.Seq - 1
-	sSeq := segment.Seq - 1
+	tSeq := t.Seq
```
Comment (cgilmour) [inline ipam/ipam.go]: Since it doesn't do arithmetic on the value of `t.Seq` and `segment.Seq` anymore, it could just use those directly instead of creating `tSeq` and `sSeq`.

File: common/defs.go (hunk 1 above)
Comment (debedb) [inline common/defs.go]: Well, the reason it produces JSON is that JSON is really easy to read (vs reading %v or %#v), and really easy to create. Compare logs where %#v is used vs JSON and you will see what is easier.

File: ipam/ipam.go (hunk 4 above)
Comment (debedb) [inline ipam/ipam.go]: Because this really should be something like Log level debug or trace...

File: ipam/ipam.go (hunk 6 above)
Comment (debedb) [inline ipam/ipam.go]: yeah

File: romana/kubernetes/listener.go [hunk 7]
```diff
@@ -367,14 +366,16 @@ func (l *kubeListener) applyNetworkPolicy(action networkPolicyAction, romanaNetw
 		return err
 	}
 	policyURL = fmt.Sprintf("%s/policies", policyURL)
+	policyStr, _ := json.Marshal(romanaNetworkPolicy)
```
Comment (cgilmour) [inline romana/kubernetes/listener.go]: The error value is ignored. Is it impossible to have an error here?

File: test/integration_test.go [hunk 8]
```diff
@@ -55,8 +56,30 @@ func myLog(c *check.C, args ...interface{}) {
...
+			newArgs[i] = a
+		case uint:
+			newArgs[i] = a
+		case uint64:
+			newArgs[i] = a
+		case string:
+			newArgs[i] = a
+		}
+	}
+	fmt.Printf(args[0].(string), newArgs...)
```
Comment (cgilmour) [inline test/integration_test.go]: This logs to stdout as well as c.Log? Is that the appropriate amount of noise?

Comment (cgilmour) [inline test/integration_test.go]: Also if the first argument must be a string (unchecked type-assertion and array indexing), why not just make it part of the signature? eg: `func myLog(c *check.C, format string, args ...interface{}) {`

Comment (debedb) [inline test/integration_test.go]: it's noisy but it's tests...

Both comments are ok but we haven't yet settled on a nice test convention for us. It's been an ad hoc exercise.
